si5351/si5351.go

```diff
@@ -0,0 +1,443 @@
+func (d *Device) ConfigureMultisynth(output uint8, pll uint8, div uint32, num uint32, denom uint32) (err error) {
+
+	// Basic validation
+	if !d.initialised {
+		return errNotInitialised
+	}
+	// Channel range
+	if !(output < 3) {
+		return errInvalidParameter
+	}
+	// Divider integer value
+	if !((div > 3) && (div < 2049)) {
+		return errInvalidParameter
+	}
+	// Avoid divide by zero
+	if !(denom > 0) {
+		return errInvalidParameter
+	}
+	// 20-bit limit
+	if !(num <= 0xFFFFF) {
+		return errInvalidParameter
+	}
+	// 20-bit limit
+	if !(denom <= 0xFFFFF) {
+		return errInvalidParameter
+	}
+
+	// Make sure the requested PLL has been initialised
+	if pll == PLL_A && !d.pllaConfigured {
+		return errInvalidParameter
+	}
+	if pll == PLL_B && !d.pllbConfigured {
+		return errInvalidParameter
+	}
+
+	// Output Multisynth Divider Equations
+	//
+	// where: a = div, b = num and c = denom
+	//
+	// P1 register is an 18-bit value using following formula:
+	//
+	//  P1[17:0] = 128 * a + floor(128*(b/c)) - 512
+	//
+	// P2 register is a 20-bit value using the following formula:
+	//
+	//  P2[19:0] = 128 * b - c * floor(128*(b/c))
+	//
+	// P3 register is a 20-bit value using the following formula:
+	//
+	//  P3[19:0] = c
+	//
+
+	// Set PLL config registers
+	var p1, p2, p3 uint32
+	if num == 0 {
+		// Integer mode
+		p1 = 128*div - 512
+		p2 = 0
+		p3 = denom
+	} else if denom == 1 {
+		// Fractional mode, simplified calculations
+		p1 = 128*div + 128*num - 512
+		p2 = 128*num - 128
+		p3 = 1
+	} else {
+		// Fractional mode
+		p1 = uint32(128*float64(div) + math.Floor(128*(float64(num)/float64(denom))) - 512)
+		p2 = uint32(128*float64(num) - float64(denom)*math.Floor(128*(float64(num)/float64(denom))))
+		p3 = denom
+	}
+
+	// Get the appropriate starting point for the PLL registers
+	baseaddr := uint8(0)
+	switch output {
+	case 0:
+		baseaddr = MULTISYNTH0_PARAMETERS_1
+		break
+	case 1:
+		baseaddr = MULTISYNTH1_PARAMETERS_1
+		break
+	case 2:
+		baseaddr = MULTISYNTH2_PARAMETERS_1
+		break
+	}
+
+	// Set the MSx config registers
+	data := d.buf[:8]
+	data[0] = uint8((p3 & 0xFF00) >> 8)
+	data[1] = uint8(p3 & 0xFF)
+	data[2] = uint8(((p1 & 0x30000) >> 16)) | d.lastRdivValue[output]
+	data[3] = uint8((p1 & 0xFF00) >> 8)
+	data[4] = uint8(p1 & 0xFF)
+	data[5] = uint8(((p3 & 0xF0000) >> 12) | ((p2 & 0xF0000) >> 16))
+	data[6] = uint8((p2 & 0xFF00) >> 8)
+	data[7] = uint8(p2 & 0xFF)
+	err = d.bus.WriteRegister(d.Address, baseaddr, data)
+	if err != nil {
+		return
+	}
+
+	// Configure the clk control and enable the output
+	// TODO: Check if the clk control byte needs to be updated.
+	clkControlReg := uint8(0x0F) // 8mA drive strength, MS0 as CLK0 source, Clock not inverted, powered up
+	if pll == PLL_B {
+		clkControlReg |= (1 << 5) // Uses PLLB
+	}
+	if num == 0 {
+		clkControlReg |= (1 << 6) // Integer mode
+	}
+
+	var register uint8
+	switch output {
+	case 0:
+		register = CLK0_CONTROL
+		break
```

Comment:
don't need the breaks.

btw, golangci-lint would likely make useful suggestions.